File: holbertonschool-higher_level_programming/discover_more_python/github_locations_concurrent.py
import json
import urllib
import urllib2
import threading
import Queue

queue = Queue.Queue()
full_result = []

def grabUrl (url):
    head = {
        'User-Agent': 'Holberton_school',
        'Authorization': 'token 1b97b1e74f551c675ad4bd92b6c327bd0fe2253d'
    }

    req = urllib2.Request(url, headers=head)
    res = urllib2.urlopen(req)
    ans = json.load(res)

    return ans

class urlThread(threading.Thread):
    def __init__(self, queue):
        threading.Thread.__init__(self)
        self.queue = queue

    def run(self):
        while True:
            d = self.queue.get()
            new_Res = grabUrl('https://api.github.com/users/' + d["owner"]["login"])

            obj = {
                "full_name": d["full_name"],
                "location": new_Res["location"]
            }

            full_result.append(obj)
            self.queue.task_done()

def main():
    url = 'https://api.github.com/search/repositories?q=language:python&sort=stars&order=desc'

    ans = grabUrl(url)

    for d in ans["items"]:
        t = urlThread(queue)
        t.setDaemon(True)
        t.start()

        queue.put(d)

    queue.join()

    print json.dumps(full_result, sort_keys=True)

main()

File: holbertonschool-higher_level_programming/discover_ruby/github_locations_concurrent.rb
require 'HTTPClient'
require 'json'

extheaders = {
  'User-Agent' => 'Holberton_School',
  'Authorization' => 'token 9f64412d7bbda257edb5a3c976db2379a43486f5'
}

uri = 'https://api.github.com/search/repositories?q=language:ruby&sort=stars&order=desc'

clnt = HTTPClient.new
res = JSON.parse(clnt.get_content(uri, extheaders))

full_res = []
req_arr = []
res["items"].map do |var|
  req_arr << Thread.new(var["owner"]["url"], extheaders) do |a, b|
    item_res = JSON.parse(clnt.get_content(a, "q=language:ruby", b))
    item_next = {
      "full_name" => var["full_name"],
      "location" => item_res["location"]
    }
    full_res.push(item_next)
  end
end

req_arr.each { |x| x.join }
puts full_res.to_json
